### PR TITLE
JP-3784: Fix off-nominal pixel_replace and cube_build output names

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3784](https://jira.stsci.edu/browse/JP-3784)

Fixes output filenames for pixel_replace and cube_build when the steps are called outside the pipelines.  For pixel_replace, only the case where the input is a ModelContainer is affected.

Based on #8979, with a few additional fixes and test updates.  

This PR also includes a minor refactor to move the `invariant_filename` function from spec3 to stpipe.utilities.  In the end, that function wasn't needed to fix this problem, but it still seems like a more useful place for that function to be stored.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by melanieclarke** [comment]
Regression tests running here:
https://github.com/spacetelescope/RegressionTests/actions/runs/12401830665

File: jwst/pixel_replace/pixel_replace_step.py [hunk 1]
```diff
@@ -72,12 +71,14 @@ def process(self, input):
                 'n_adjacent_cols': self.n_adjacent_cols,
             }
 
-            # ___________________
-            # calewbb_spec3 case
-            # ___________________
+            # ___________________________________
+            # calewbb_spec3 case / ModelContainer
```
Comment (emolter) [inline jwst/pixel_replace/pixel_replace_step.py:75]:
```suggestion
            # calwebb_spec3 case / ModelContainer
```

File: jwst/pixel_replace/pixel_replace_step.py [hunk 2]
```diff
@@ -72,12 +71,14 @@ def process(self, input):
...
+            # __________________________________
             if isinstance(input_model, datamodels.ModelContainer):
                 output_model = input_model
-                # Setup output path naming if associations are involved.
+
+                # Setup output path naming if associations are involved, to
+                # include the ASN ID in the output
```
Comment (emolter) [inline jwst/pixel_replace/pixel_replace_step.py:81]:
```suggestion
                # Set up output path name to include the ASN ID if associations are involved
```

**Review by emolter — APPROVED** [review]
LGTM, I like the improvements to `test_invariant_filename`.  Admittedly I only skimmed this because I assume it is similar to Jane's PR that I already looked at

**Comment by melanieclarke** [comment]
@emolter - this is the updated fix for the pixel_replace output names.  I will leave it at draft so it doesn't get merged for 11.2, but it is ready for review.

It turns out the only fix necessary was to default the pixel replace step to output_use_model=True.  This is done for several other similar steps where multiple input are used, but output files should be based on the input name (tweakreg, master_background, cube_build), and should probably be added anywhere we might reasonably expect the input and output to be a list of files.

I kept the change to put invariant_filename into stpipe.utilities, because that seems like the right place for it anyway, but it is no longer needed to fix the filenames.

**Comment by emolter** [comment]
> It turns out the only fix necessary was to default the pixel replace step to output_use_model=True. 

So to be clear, this is an stpipe option available to all steps, and it's added to the spec just because we want to change the default value?

It seems to me we might reasonably expect a list of files in any step that has a container as output.  Are you thinking of any other steps in particular?

**Comment by melanieclarke** [comment]
Now that 1.17.0 is released, I will merge this when CI completes.

**Comment by jemorrison** [comment]
@melanieclarke Thank you for getting this merged.

**Pull request merged**